
# ccfind - Circular Complete sequence FINDer

## description
* `ccfind` is a general tool to find circular complete sequences by detection of terminal redundancy. Originally, this tool was intended for identification of complete virus genomes.
* `ccfind` can be used for any contig/genome, but cutoff values (see `algorithm` section) should be carefully considered.

## requirements
* ssearch (included in [FASTA package](http://fasta.bioch.virginia.edu/fasta_www2/fasta_list2.shtml))
* BLAST+ 
* Ruby (ver >=2.0)
* GNU parallel (if you need parallel computing)

## algorithm
1. extract the first and last `N` bp (`N` is specified by `--terminal-fragment-size`) for each input sequence. Note that if sequence length is less than `N` x 2, the sequence will be ignored.
2. detect sequence homology between the two fragments by `blastn` (as screening).
3. For pairs that passed the screening, compute alignment between them by `ssearch`.
4. If the alignment is not begin with the start of the first part, extend alignments to the start, with regarding the exended region as mismatch.
5. If the alignment is not begin with the end of the last part, extend alignments to the end, with regarding the exended region as mismatch.
6. extract 'good' alignment defined by cutoff values (`--min-percent-identity` and `--min-aligned-length`).

## usage 
```
### ccfind ver 1.2.0 (2018-07-17) ###

[description]
ccfind is tool to find circular complete sequence by detecting terminal redundancy.

ccfind is originally designed to find circular complete contigs generated by the SPAdes assembler.
ccfind can be used for any contig/genome, but cutoff values (see below) should be carefully considered.

[dependencies]
- ssearch        -- included in FASTA program
                    (available here: http://fasta.bioch.virginia.edu/fasta_www2/fasta_list2.shtml)
- blastn         -- included in the BLAST+ program;
                    (available here: https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download)
- ruby (ver >=2.0)

[usage]
$ ccfind <input fasta> <output dir> [options]

- <output dir> should not exist.

[options]
  (general)
    -h, --help
    -v, --version

  (cutoff values)
    -L, --terminal-fragment-size  [int] (default: 500) -- This should be longer than estimated terminal redundancy length. If input length is less than 2 x (this value), the sequence will be ignored and listed in "result/01.skipped.list".
    -i, --min-percent-identity    [int] (default: 94)
    -l, --min-aligned-length      [int] (default: 50)

  (use GNU parallel)
    --ncpus        [int]    -- number of jobs in parallel

  (for icr user)            -- for computation in the ICR supercomputer system
    --queue        [JP1]    -- queue for computation

[output files]
result/05.circ.detected.out -- list of sequence with terminal redundancy.
result/05.circ.fasta        -- fasta file of sequences with terminal redundancy.
result/05.circ.noTR.fasta   -- fasta file of sequences with terminal redundancy, but the detected terminal redundnacy at the end of sequence is removed.
```

## citation
If you use results genereted by ccfind in your research, please cite:
```
Environmental Viral Genomes Shed New Light on Virus-Host Interactions in the Ocean.
mSphere 2(2):e00359-16 (2017), doi:10.1128/mSphere.00359-16
Yosuke Nishimura, Hiroyasu Watai, Takashi Honda, Tomoko Mihara, Kimiho Omae, Simon Roux, Romain Blanc-Mathieu, Keigo Yamamoto, Pascal Hingamp, Yoshihiko Sako, Matthew B. Sullivan, Susumu Goto, Hiroyuki Ogata, Takashi Yoshida
```
